IpmiFeaturePkg/Library/IpmiTransportLibMock/MockIpmi.c

```diff
@@ -0,0 +1,181 @@
+/** @file
+  A mock implementation of a BMC. This allow of minimal use of the IPMI
+  functionality without an actual BMC. This is intended for use in testing
+  changes to the IPMI feature package as well as components that depend on it.
+
+  Copyright (c) Microsoft Corporation
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include "MockIpmi.h"
+
+BOOLEAN        gResponseValid;
+IPMI_RESPONSE  gResponse;
+UINT8          gResponseSize;
+
+/**
+  Mocks the result of IPMI_APP_GET_SELFTEST_RESULTS.
+
+  @param[in]       Data           The IPMI request data.
+  @param[in]       DataSize       The size of the IPMI request data.
+  @param[out]      Response       The response data buffer.
+  @param[in, out]  ResponseSize   On input, the available size of buffer.
+                                  On output, the size of written data in the buffer.
+**/
+VOID
+MockIpmiGetSelfTest (
+  IN VOID       *Data,
+  IN UINT8      DataSize,
+  OUT VOID      *Response,
+  IN OUT UINT8  *ResponseSize
+  )
+{
+  IPMI_SELF_TEST_RESULT_RESPONSE  *SelfTest;
+
+  ASSERT (*ResponseSize >= sizeof (IPMI_SELF_TEST_RESULT_RESPONSE));
+
+  SelfTest                 = Response;
+  SelfTest->CompletionCode = IPMI_COMP_CODE_NORMAL;
+  SelfTest->Result         = IPMI_APP_SELFTEST_NOT_IMPLEMENTED;
+  SelfTest->Param          = 0;
+
+  *ResponseSize = sizeof (IPMI_SELF_TEST_RESULT_RESPONSE);
+}
+
+/**
+  Mocks the result of IPMI_APP_GET_DEVICE_ID.
+
+  @param[in]       Data           The IPMI request data.
+  @param[in]       DataSize       The size of the IPMI request data.
+  @param[out]      Response       The response data buffer.
+  @param[in, out]  ResponseSize   On input, the available size of buffer.
+                                  On output, the size of written data in the buffer.
+**/
+VOID
+MockIpmiGetDeviceId (
+  IN VOID       *Data,
+  IN UINT8      DataSize,
+  OUT VOID      *Response,
+  IN OUT UINT8  *ResponseSize
+  )
+{
+  IPMI_GET_DEVICE_ID_RESPONSE  *DeviceId;
+
+  ASSERT (*ResponseSize >= sizeof (IPMI_GET_DEVICE_ID_RESPONSE));

```

Comment:
Is there concern where the incoming pointers might be NULL? Should we check those? (and other places in this file as well?)

Reply:
Oops, this check is inverted. This should be checking that the buffer is large enough.

Reply:
For NULL, i'll probably do those checks at the top level function to minimize input checking expected on the handlers to reduce duplicate code.